src/www/app/outline_server.ts

```diff
@@ -88,10 +89,244 @@ export class OutlineServer implements PersistentServer {
   }
 
   checkReachable(): Promise<boolean> {
-    return this.tunnel.isReachable();
+    return this.tunnel.isReachable(this.config);
   }
 
-  public static isServerCipherSupported(cipher?: string) {
+  static isServerCipherSupported(cipher?: string) {
     return cipher !== undefined && OutlineServer.SUPPORTED_CIPHERS.includes(cipher);
   }
 }
+
+// Persistence format for an Outline Server.
+interface OutlineServerJson {
+  // Shadowsocks access key encoding proxy configuration parameters.
+  readonly accessKey: string;
+  // User-given name.
+  readonly name: string;
+}
+
+export interface ConfigByIdV0 {
+  [serverId: string]: ShadowsocksConfig;
+}
+
+export interface ConfigById {
+  [serverId: string]: OutlineServerJson;
+}
+
+export type OutlineServerFactory =
+    (id: string, name: string, config: ShadowsocksConfig, eventQueue: events.EventQueue) =>
+        OutlineServer;
+
+// Maintains a persisted set of servers and liaises with the core.
+export class OutlineServerRepository implements ServerRepository {
+  // Name by which servers are saved to storage.
+  public static readonly SERVERS_STORAGE_KEY_V0 = 'servers';
+  public static readonly SERVERS_STORAGE_KEY = 'servers_v1';
+  private serverById!: Map<string, OutlineServer>;
+  private lastForgottenServer: OutlineServer|null = null;
+
+  constructor(
+      public readonly createServer: OutlineServerFactory, private eventQueue: events.EventQueue,
+      private storage: Storage) {
+    try {
+      migrateServerStorageToV1(this.storage);
+    } catch (e) {
+      console.error(`failed to migrate storage to V1: ${e}`);
+      return;
+    }
+    this.loadServers();
+  }
+
+  getAll() {
+    return Array.from(this.serverById.values());
+  }
+
+  getById(serverId: string) {
+    return this.serverById.get(serverId);
+  }
+
+  add(accessKey: string, serverName: string) {
```

Comment:
Let's make add not take the serverName. I'd like this to be aligned with the product.
It's ok to have unnamed servers. The application/ui code should have the logic: if server has a name, use it, otherwise use the default name. This is the same approach we use for access keys in the manager.

In order to support Outline/Shadowsocks Server, you'll need to expose a getter to differentiate them. 

Note that the access key may have a name, so we can leverage that for the fake servers.

Reply:
Done in 76bde1c.  I also changed the persistence format to encode the name. In order to support default (outlne/non-outline) server name, I added an `extra` field to `ShadowsocksConfig`.